### ISSUE TITLE
specshow: 'log' frequency scale is not really logarithmic

### ISSUE THREAD

**Issue by sebewert** [librosa/librosa#329]
Hi,
it seems the axis labels shown using specshow with y_axis='log' do not seem to be really logarithmic. For example, using the code snippet below I get a plot with the following equidistant labels: 0, 829.03, 2002.6, 4005.2, 22050

Unless I am confused right now, this cannot be logarithmic. In particular, log(2 x) = log(2) + log(x)
That means whenever one value is twice as big as the next one, their difference on the logarithmic scale will be log(2). This is independent of the logarithm base in use. 

The distance on the logarithmic scale is supposed to be equidistant four all values above. One pair is particularly interesting as  2002.6 * 2 = 4005.2. That means the distance between them gives log(2). The next label above 4005.2, having the same distance on the logarithmic scale, is 22050 and not 8010.4. Unless I misunderstand the docs 'log : the image is displayed on a vertical log scale', that means there might be an issue here somehow.

```python
from __future__ import ( division, absolute_import, print_function, unicode_literals )
import librosa
import matplotlib.pyplot as plt
import numpy as np

sr=44100
y = np.random.rand(sr*2)

n_fft = 4096
hop_length = n_fft/4
Z = librosa.core.stft(y, n_fft=n_fft, hop_length=hop_length, center=True)

plt.figure(figsize=(8,6))
librosa.display.specshow(np.abs(Z), sr=sr, hop_length=hop_length, x_axis='time', y_axis='log')
```

**Comment by bmcfee** [comment]
I think I see what's going on here, and it's a little complicated.  (log-scale display is always more of a pain than it ought to be. :)) For reference, the axis scale positioning code is [here](https://github.com/bmcfee/librosa/blob/master/librosa/display.py#L995).

The short answer here is that the positioning isn't really mapping `f` to `log2(f)`, but is instead dividing the range of FFT bins into exponentially bigger chunks.

The first and last ticks are always at 0 (DC) and nyquist; the ticks in between will be log-scaled.  The 0-tick obviously doesn't exist in log-frequency, but we draw the tick anyway.

If you say `n_yticks=10` in `specshow`, maybe the behavior will be more obvious.

Of course, I'm open to suggestions on how to do this part better.

**Comment by sebewert** [comment]
I had a look at the code. I gave it a go but halfway through it became obvious that I didn't really have a clue what it really is one wants from this. It really is a bit tricky...

**Comment by bmcfee** [comment]
> it became obvious that I didn't really have a clue what it really is one wants from this.

Yup, I came to the same conclusion back when this was first implemented!  I think log-scaling the bin widths (rather than frequencies) is really the only sensible thing to do, but it's hard to do this with consistently placed tick marks that also match up at the edges.

In general, log-axis specshow is a big pile of hack that should probably be rewritten by someone more clever than myself.  (The double-drawing is particularly bad, but it works.)

**Comment by bmcfee** [comment]
@sebewert Following up on this: could you take a peek at #365 , and specifically, the new test images (viewable starting [here](https://github.com/librosa/librosa/pull/368/files#diff-0086c8faaddf0e94697df445fb8d7940))?

TLDR: total rewrite of the plot code so that it now uses matplotlib's log and symlog axis scaling.  The code is now much simpler, since we no longer do any explicit coordinate mapping.

For [log-frequency fft](https://github.com/librosa/librosa/pull/368/files#diff-8c9b43204af8c4292a5b1dd5539390ed) display, it now uses a linear axis up to C2 and log2 scaling after that.  The linear region (0-C2) is stretched to cover half the range of the remaining octaves.  The results are not quite the same as the previous pseudo-log scaling discussed in this issue, so I'd like some feedback before merging.

For [mel](https://github.com/librosa/librosa/pull/368/files#diff-3c0d08439054734573a198b5f9fd1591) display. it uses linear frequencies up to 1KHz and log2 after that, which should be consistent with Slaney-style mel filter construction.  (Htk is not that far off, but not exactly linear in the low end.)

For [tempo](https://github.com/librosa/librosa/pull/368/files#diff-6f45edba3ab16f5d91b46d4f10747161) and [CQT-style](https://github.com/librosa/librosa/pull/368/files#diff-0086c8faaddf0e94697df445fb8d7940) displays, frequency is log2 over the entire axis.

In all cases, the frequency units are now in Hz (or BPM for tempo) regardless of axis scaling.  I think this should resolve your initial issue here, but confirmation would be appreciated.

**Comment by bmcfee** [comment]
Resolved by merging #368
